xarray/core/dataset.py

```diff
@@ -1485,11 +1496,20 @@ def __setitem__(self, key: List[Hashable], value) -> None:
     def __setitem__(self, key: Hashable, value) -> None:
         ...
 
+    @overload
+    def __setitem__(self, key: Any, value) -> None:
```

Comment:
Doesn't the new case need to be a `Mapping`? If so, `Any` will work but is overly permissive

Reply:
Yeah I wasn't sure there. If I use `Mapping` instead of `Any`, then I get the following hook error from mypy:
`xarray/core/dataset.py:1498: error: Overloaded function signature 3 will never be matched: signature 2's parameter type(s) are the same or broader`
Can we just delete this overloaded function signature?

Reply:
@max-sixty what do you think?

Reply:
Thanks for the ping!

I get the same error. I don't understand it, though, since `Mapping` isn't necessarily a `Hashable`, which is I think is the 2 & 3 it refers to.

That said, `Any` is overly wide and I think not correct.

Surprisingly, this works, and doesn't lose any specificity since they all have the same return type:

```diff
diff --git a/xarray/core/dataset.py b/xarray/core/dataset.py
index a96c2b12..3880ab24 100644
--- a/xarray/core/dataset.py
+++ b/xarray/core/dataset.py
@@ -1488,19 +1488,7 @@ def __getitem__(self, key):
         else:
             return self._copy_listed(np.asarray(key))
 
-    @overload
-    def __setitem__(self, key: List[Hashable], value) -> None:
-        ...
-
-    @overload
-    def __setitem__(self, key: Hashable, value) -> None:
-        ...
-
-    @overload
-    def __setitem__(self, key: Any, value) -> None:
-        ...
-
-    def __setitem__(self, key, value) -> None:
+    def __setitem__(self, key: Union[Hashable, List[Hashable], Mapping], value) -> None:
         """Add an array to this dataset.
         Multiple arrays can be added at the same time, in which case each of
         the following operations is applied to the respective value.

```

Reply:
that's strange, but ok I made the suggested change.